File: Leetcode/SearchinRotatedSortedArray.java
class Solution {
    public int search(int[] nums, int target){
        int pivot = findPivot(nums);

        // if you don't find a pivot, it means the array is not rotated
        if(pivot == -1){
            // just do normal binary search
            return binarySearch(nums,target,0,nums.length - 1);
        }

        // if pivot is found, you have 2 asc sorted arrays
        if(nums[pivot] == target){
            return pivot;
        }
        if( target >= nums[0]){
            return binarySearch(nums,target,0,pivot - 1);
        }

        return binarySearch(nums,target,pivot + 1,nums.length - 1);
    }

     int binarySearch(int[] arr, int target,int start, int end){

        while(start<=end){

            int mid = start + (end - start) / 2;

            if (target < arr[mid]){
                end = mid - 1;
            } else if (target > arr[mid]){
                start = mid +1;
            } else {
                //ans found
                return mid;
            }
        }
        return -1;
    }

     int findPivot(int[] arr){
        int start = 0;
        int end = arr.length - 1;
        while(start <=end){
            int mid = (end - start) / 2 + start ;

            // 4 cases here
            if(mid < end && arr[mid] > arr[mid + 1]){
                return mid;
            }
            if(mid > start && arr[mid] < arr[mid - 1]){
                return mid - 1;
            }
            if(arr[mid] <= arr[start]){
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return -1;
    }
}

File: Leetcode/Search in Rotated Sorted Array.cpp
#include <bits/stdc++.h>
using namespace std;

// Problem Link: https://leetcode.com/problems/search-in-rotated-sorted-array/

class Solution {
    int findMin(vector<int>& nums) {
        if (nums.size() == 1) return 0;
        if (nums[nums.size()-1] > nums[0]) return 0;
        
        int l=0,r=nums.size()-1,mid;
        while (l<r) {
            mid = (l+r)/2;
            if (mid > 0 && nums[mid] < nums[mid-1]) return mid;
            if (mid < nums.size() - 1 && nums[mid+1] < nums[mid]) return mid+1;
            
            if (nums[mid] > nums[0]) l = mid+1;
            else r = mid-1;
        }
        
        return l;
    }
    int binSearch(vector<int>& nums,int l, int r, int target) {
        while (l<=r) {
            int mid = (l+r)/2;
            if (target == nums[mid]) return mid;
            if (target > nums[mid]) l = mid+1;
            else r = mid-1;
        }
        return -1;
    }
public:
    int search(vector<int>& nums, int target) {
        int min = findMin(nums);
        if (target < nums[min] || (min > 0 && target > nums[min-1])) return -1;

        if (min == 0) return binSearch(nums,0,nums.size()-1,target);
        if (target >= nums[0]) return binSearch(nums,0,min-1,target);
        return binSearch(nums,min,nums.size()-1,target);
    }
};